rfd/0176/README.md

```diff
@@ -66,48 +71,64 @@ There are a couple dimensions to each of those:
 
 In this scenario, the boot process is:
 
+#### Original proposal
+(untested thus far)
 1. Hardware loads master boot record (MBR) from a boot disk
 2. The code in the MBR loads iPXE from the ZFS boot area on that same disk
 3. iPXE performs a network boot
 
-### Local boot with BIOS
+#### Modified proposal
+(tested successfully with Triton CN)
+1. Hardware loads master boot record (MBR) from a boot disk
+2. The code in the MBR executes the illumos loader from the ZFS boot
```

Comment:
The MBR is 512 bytes.  You seem to be saying that in this 512 bytes there is:

- MBR magic number so it is recognized as an MBR by the BIOS.
- Code that can find pool properties then interpret that to find bootfs.  Presumably this step could be simplified by storing the LBA where the pool configuration is stored (as a packed nvlist, if memory serves).
- Code that can read the pool configuration, find the dataset referred to by `bootfs`, then issue the appropriate I/Os even in the face of raidz and compression.
- Code that can then understand enough of the ZFS POSIX layer to find and read a file in that dataset.

Maybe that's possible in 512 bytes.  I have a hunch that something is placing loader in a well-known location in sequential sectors.  The MBR is loading loader from that location then loader is finding configuration and perhaps additional code in bootfs.

Reply:
Please see https://github.com/joyent/smartos-live/blob/master/tools/format_image/format_image.c#L14

Reply:
Editing that section.